pkg/mcp/tests/stress/stress_test.go

```diff
@@ -0,0 +1,634 @@
+	nackRate             = flag.Float64("nack_rate", 0.01, "rate at which clients nack updates. Should be in the range [0,1]")
+	clientIncPercentage  = flag.Float64("client_inc_rate", 0, "percentate of clients that request incremental updates [0,1]")
+)
+
+var (
+	commonDataset []snapshot.Snapshot
+)
+
+func TestMain(m *testing.M) {
+	flag.Parse()
+
+	// reduce logging from grpc library
+	var (
+		infoW    = ioutil.Discard
+		warningW = ioutil.Discard
+		errorW   = os.Stderr
+	)
+	grpclog.SetLoggerV2(grpclog.NewLoggerV2(infoW, warningW, errorW))
+
+	// reduce logging from MCP library
+	o := log.DefaultOptions()
+	o.SetOutputLevel("mcp", log.NoneLevel)
+	o.SetOutputLevel("default", log.NoneLevel)
+	log.Configure(o)
+
+	fmt.Println("Generating common dataset ...")
+	initDataset()
+	fmt.Println("Finished generating common dataset")
+
+	// call flag.Parse() here if TestMain uses flags
+	os.Exit(m.Run())
+}
+
+func stressTestCommon(tb testing.TB, o options) {
+	tb.Helper()
+
+	d, err := newDriver(o)
+	if err != nil {
+		tb.Fatalf("error creating driver: %v", err)
+	}
+	defer d.Close()
+	d.run(tb)
+}
+
+func TestFullState(t *testing.T) {
+	o := defaultOptions()
+	stressTestCommon(t, o)
+}
+
+func TestServerSupportedIncrementalOnly(t *testing.T) {
+	o := defaultOptions()
+	o.serverIncemental = true
+	stressTestCommon(t, o)
+}
+
+func TestServerSupportedIncremental50PercentClients(t *testing.T) {
+	o := defaultOptions()
+	o.serverIncemental = true
+	o.clientIncRate = 0.5
+	stressTestCommon(t, o)
+}
+
+func TestServerSupportedIncremental100PercentClients(t *testing.T) {
+	o := defaultOptions()
+	o.serverIncemental = true
+	o.clientIncRate = 1
+	stressTestCommon(t, o)
+}
+
+func TestFullStateSlowClients(t *testing.T) {
+	o := defaultOptions()
+	o.serverIncemental = true
+	o.clientIncRate = 1
+	o.minApplyDelay = 50 * time.Millisecond
+	o.maxApplyDelay = 100 * time.Millisecond
+	stressTestCommon(t, o)
+}
+
+// NOTE: these benchmarks don't provide meaningful ns/ops numbers in the
+// traditional sense. These are intended to be used with the -cpuprofile and -memprofile
+// options along with `go tool pprof` to analyze resource usage in various cases.
+
+func BenchmarkFullState(b *testing.B) {
+	o := defaultOptions()
+	o.iterations = b.N
+	o.numClients = 50
+	stressTestCommon(b, o)
+}
+
+func BenchmarkServerSupportedIncrementalOnly(b *testing.B) {
+	o := defaultOptions()
+	o.iterations = b.N
+	o.numClients = 50
+	o.serverIncemental = true
+	stressTestCommon(b, o)
+}
+
+func BenchmarkServerSupportedIncremental50PercentClients(b *testing.B) {
+	o := defaultOptions()
+	o.iterations = b.N
+	o.numClients = 50
+	o.serverIncemental = true
+	o.clientIncRate = 0.5
+	stressTestCommon(b, o)
+}
+
+func BenchmarkServerSupportedIncremental100PercentClients(b *testing.B) {
+	o := defaultOptions()
+	o.iterations = b.N
+	o.numClients = 50
+	o.serverIncemental = true
```

Comment:
I assume serverIncremental implies incremental as the behavior is controlled by server at this point? Can you add comments for this?

Reply:
Renamed to serverIncSupported and aligned with the flag naming